pkg/controllermanager/controller/project/resourcequota/reconciler.go

```diff
@@ -0,0 +1,96 @@
+// SPDX-FileCopyrightText: SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package resourcequota
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	controllermanagerconfigv1alpha1 "github.com/gardener/gardener/pkg/controllermanager/apis/config/v1alpha1"
+	gardenerutils "github.com/gardener/gardener/pkg/utils/gardener"
+)
+
+const shootCountResource = "count/shoots.core.gardener.cloud"
+
+var gardenerCreatedResourcesCounts = map[corev1.ResourceName]int{
+	"count/configmaps": len(gardenerutils.GetShootProjectConfigMapSuffixes()),
+	"count/secrets":    len(gardenerutils.GetShootProjectSecretSuffixes()),
+}
+
+// Reconciler reconciles ResourceQuotas in Project namespaces.
+// It ensures that ResourceQuota objects in project namespaces allow enough resources for all possible Shoot resources.
+// It checks the current quota limits and adjusts them if needed, so that the namespace can accommodate the maximum number of Shoots and their related resources (like ConfigMaps and Secrets) according to the configured limits.
+type Reconciler struct {
+	Client client.Client
+	Config controllermanagerconfigv1alpha1.ProjectControllerConfiguration
+}
+
+// Reconcile adjusts the ResourceQuota in a Project namespace to ensure that it can accommodate all Shoots, according to the Shoot limit.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(ctx)
+
+	resourceQuota := &corev1.ResourceQuota{}
+	if err := r.Client.Get(ctx, request.NamespacedName, resourceQuota); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	if resourceQuota.DeletionTimestamp != nil || resourceQuota.Spec.Hard == nil {
+		return reconcile.Result{}, nil
+	}
+
+	patch := client.MergeFrom(resourceQuota.DeepCopy())
```

Comment:
We _could_ think about moving the `patch` var specification down to `L71`, but not a must. 😉